File: bindings/cxx/tests/tests-chip-info.cpp
// SPDX-License-Identifier: GPL-2.0-or-later
// SPDX-FileCopyrightText: 2021-2022 Bartosz Golaszewski <brgl@bgdev.pl>

#include <catch2/catch_all.hpp>
#include <gpiod.hpp>
#include <sstream>

#include "gpiosim.hpp"
#include "helpers.hpp"

using ::gpiosim::make_sim;

namespace {

TEST_CASE("chip_info properties can be read", "[chip-info][chip]")
{
	auto sim = make_sim()
		.set_num_lines(8)
		.set_label("foobar")
		.build();

	::gpiod::chip chip(sim.dev_path());
	auto info = chip.get_info();

	SECTION("get chip name")
	{
		REQUIRE_THAT(info.name(), Catch::Matchers::Equals(sim.name()));
	}

	SECTION("get chip label")
	{
		REQUIRE_THAT(info.label(), Catch::Matchers::Equals("foobar"));
	}

	SECTION("get num_lines")
	{
		REQUIRE(info.num_lines() == 8);
	}
}

TEST_CASE("chip_info can be copied and moved", "[chip-info]")
{
	auto sim = make_sim()
		.set_num_lines(4)
		.set_label("foobar")
		.build();

	::gpiod::chip chip(sim.dev_path());
	auto info = chip.get_info();

	SECTION("copy constructor works")
	{
		auto copy(info);

		REQUIRE_THAT(copy.name(), Catch::Matchers::Equals(sim.name()));
		REQUIRE_THAT(copy.label(), Catch::Matchers::Equals("foobar"));
		REQUIRE(copy.num_lines() == 4);

		REQUIRE_THAT(info.name(), Catch::Matchers::Equals(sim.name()));
		REQUIRE_THAT(info.label(), Catch::Matchers::Equals("foobar"));
		REQUIRE(info.num_lines() == 4);
	}

	SECTION("assignment operator works")
	{
		auto copy = chip.get_info();

		copy = info;

		REQUIRE_THAT(copy.name(), Catch::Matchers::Equals(sim.name()));
		REQUIRE_THAT(copy.label(), Catch::Matchers::Equals("foobar"));
		REQUIRE(copy.num_lines() == 4);

		REQUIRE_THAT(info.name(), Catch::Matchers::Equals(sim.name()));
		REQUIRE_THAT(info.label(), Catch::Matchers::Equals("foobar"));
		REQUIRE(info.num_lines() == 4);
	}

	SECTION("move constructor works")
	{
		auto moved(std::move(info));

		REQUIRE_THAT(moved.name(), Catch::Matchers::Equals(sim.name()));
		REQUIRE_THAT(moved.label(), Catch::Matchers::Equals("foobar"));
		REQUIRE(moved.num_lines() == 4);
	}

	SECTION("move assignment operator works")
	{
		auto moved = chip.get_info();

		moved = ::std::move(info);

		REQUIRE_THAT(moved.name(), Catch::Matchers::Equals(sim.name()));
		REQUIRE_THAT(moved.label(), Catch::Matchers::Equals("foobar"));
		REQUIRE(moved.num_lines() == 4);
	}
}

TEST_CASE("stream insertion operator works for chip_info", "[chip-info]")
{
	auto sim = make_sim()
		.set_num_lines(4)
		.set_label("foobar")
		.build();

	::gpiod::chip chip(sim.dev_path());
	auto info = chip.get_info();
	::std::stringstream expected;

	expected << "gpiod::chip_info(name=\"" << sim.name() <<
		    "\", label=\"foobar\", num_lines=4)";

	REQUIRE_THAT(info, stringify_matcher<::gpiod::chip_info>(expected.str()));
}

} /* namespace */

File: bindings/python/tests/tests_chip_info.py
# SPDX-License-Identifier: GPL-2.0-or-later
# SPDX-FileCopyrightText: 2022 Bartosz Golaszewski <brgl@bgdev.pl>

from unittest import TestCase

import gpiod

from . import gpiosim


class ChipInfoProperties(TestCase):
    def setUp(self) -> None:
        self.sim = gpiosim.Chip(label="foobar", num_lines=16)
        self.chip = gpiod.Chip(self.sim.dev_path)
        self.info = self.chip.get_info()

    def tearDown(self) -> None:
        self.info = None
        self.chip.close()
        self.chip = None
        self.sim = None

    def test_chip_info_name(self) -> None:
        self.assertEqual(self.info.name, self.sim.name)

    def test_chip_info_label(self) -> None:
        self.assertEqual(self.info.label, "foobar")

    def test_chip_info_num_lines(self) -> None:
        self.assertEqual(self.info.num_lines, 16)

    def test_chip_info_properties_are_immutable(self) -> None:
        with self.assertRaises(AttributeError):
            self.info.name = "foobar"

        with self.assertRaises(AttributeError):
            self.info.num_lines = 4

        with self.assertRaises(AttributeError):
            self.info.label = "foobar"


class ChipInfoStringRepresentation(TestCase):
    def test_chip_info_str(self) -> None:
        sim = gpiosim.Chip(label="foobar", num_lines=16)

        with gpiod.Chip(sim.dev_path) as chip:
            info = chip.get_info()

            self.assertEqual(
                str(info),
                '<ChipInfo name="{}" label="foobar" num_lines=16>'.format(sim.name),
            )
